hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java

```diff
@@ -157,13 +159,17 @@ protected void onInitialize() {
     public void setCacheManager(HazelcastCacheManager cacheManager) {
         assert cacheManager instanceof HazelcastClientCacheManager;
 
-        this.cacheManager = (HazelcastClientCacheManager) cacheManager;
+        if (!cacheManagerRef.compareAndSet(null, (HazelcastClientCacheManager) cacheManager)) {
+            if (cacheManagerRef.get() != cacheManager) {
+                throw new IllegalStateException("Cannot overwrite a Cache's CacheManager.");
+            }
+        }
     }
 
     @Override
     protected void postDestroy() {
-        if (cacheManager != null) {
-            cacheManager.destroyCache(getName());
+        if (cacheManagerRef.get() != null) {
```

Comment:
I normally prefer the following approach:

```
CacheManager cacheManger = cacheManagerRef.get();
if (cacheManger != null) {
	cacheManger.destroyCache(getName());
}
```

In the existing code I don't see a change in the cacheManagerRef after it has been set, but the field is protected so I don't know what is happening outside.

If you use the above approach you are taking an action based on a snapshot of reality, instead of needing to deal with world that can change underneath your feet (it could in theory lead to a NPE for example if the ref gets cleared).